maze_dataset/tokenization/maze_tokenizer.py

```diff
@@ -388,3 +429,723 @@ def clear_cache(self):
+                (
+                    target_tokenizer.to_tokens(target, coord_tokenizer=coord_tokenizer)
+                    if target[0] is not None
+                    else []
+                ),
+                (
+                    path_tokenizer.to_tokens(path, coord_tokenizer=coord_tokenizer)
+                    if path is not None
+                    else []
+                ),
+            ]
+
+        @abc.abstractmethod
+        def _sequence_tokens(
+            self,
+            adj_list: list[str],
+            origin: list[str] | None,
+            target: list[str] | None,
+            path: list[str] | None,
+        ) -> list[str]:
+            """Sequences token regions into a complete prompt.
+            Includes any boundary tokens in `constants.SPECIAL_TOKENS` such as <ADJLIST_START>, <ORIGIN_END>, etc.
+            # Parameters
+            - `adj_list`: Tokens representing the adjacency list
+            - `origin`: Tokens representing the origin
+            - `target`: Tokens representing the target
+            - `path`: Tokens representing the path
+            """
+            pass
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class AOTP(PromptSequencer):
+        def _sequence_tokens(
+            self,
+            adj_list: list[str],
+            origin: list[str],
+            target: list[str],
+            path: list[str],
+        ) -> list[str]:
+            return [
+                VOCAB.ADJLIST_START,
+                *adj_list,
+                VOCAB.ADJLIST_END,
+                VOCAB.ORIGIN_START,
+                *origin,
+                VOCAB.ORIGIN_END,
+                VOCAB.TARGET_START,
+                *target,
+                VOCAB.TARGET_END,
+                VOCAB.PATH_START,
+                *path,
+                VOCAB.PATH_END,
+            ]
+
+    @serializable_dataclass(frozen=True, kw_only=True)
+    class AOP(PromptSequencer):
+        include_target_special_tokens: bool = serializable_field(default=True)
+
+        def _sequence_tokens(
+            self,
+            adj_list: list[str],
+            origin: list[str],
+            target: list[str],
+            path: list[str],
+        ) -> list[str]:
+            return [
+                VOCAB.ADJLIST_START,
+                *adj_list,
+                VOCAB.ADJLIST_END,
+                VOCAB.ORIGIN_START,
+                *origin,
+                VOCAB.ORIGIN_END,
+                *(
+                    [VOCAB.TARGET_START, VOCAB.TARGET_END]
+                    if self.include_target_special_tokens
+                    else ()
+                ),
+                VOCAB.PATH_START,
+                *path,
+                VOCAB.PATH_END,
+            ]
+
+
+def _load_tokenizer_element(
+    data: dict[str, Any], namespace: type[_TokenizerElementNamespace]
+) -> TokenizerElement:
+    key: str = namespace.key
+    format: str = data[key]["__format__"]
+    cls_name: str = format.split("(")[0]
```

Comment:
this feels messy, but I'm not sure of a better way

Reply:
Updated in #37. Will review this method then